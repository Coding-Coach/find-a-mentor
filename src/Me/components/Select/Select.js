import PropTypes from 'prop-types';
import React, { useContext, useLayoutEffect, useRef } from 'react';
import ReactSelect from 'react-select';
import { formFieldContext } from '../FormField/formContext';

const REACHED_MAX_OPTION = [
  {
    label: 'Reached max items',
    value: undefined,
    isDisabled: true,
  },
];

export const Select = ({
  isMulti = false,
  value,
  onChange,
  maxSelections = Infinity,
  options,
  ...reactSelectOptions
}) => {
  const id = useContext(formFieldContext);
  const inputElRef = useRef(null);

  useLayoutEffect(() => {
    inputElRef.current.select.inputRef.autocomplete = 'nope';
  }, []);

  return (
    <ReactSelect
      ref={inputElRef}
      isMulti={isMulti}
      menuPlacement="auto"
      hideSelectedOptions
      inputId={id}
      isSearchable
      value={value}
      onChange={onChange}
      options={
        isMulti && value && value.length && value.length >= maxSelections
          ? REACHED_MAX_OPTION
          : options
      }
      menuPortalTarget={document.body}
      styles={reactSelectStyles}
      {...reactSelectOptions}
    />
  );
};

const selectValueShape = PropTypes.shape({
  label: PropTypes.string.isRequired,
  value: PropTypes.any.isRequired,
  isDisabled: PropTypes.bool,
});

Select.propTypes = {
  isMulti: PropTypes.bool,
  value: PropTypes.any,
  onChange: PropTypes.func.isRequired,
  maxSelections: PropTypes.number,
  options: PropTypes.arrayOf(selectValueShape),
};

const reactSelectStyles = {
  container: base => ({ ...base, width: '100%' }),
  control: base => ({
    ...base,
    fontSize: '14px',
    lineHeight: '17px',
    borderRadius: '3px',
    backgroundColor: '#fff',
    border: '1px solid #bfbfbf',
  }),
  placeholder: base => ({ ...base, color: '#898889' }),
  menuPortal: base => ({ ...base, zIndex: 1000 }),
  option: (base, { isFocused, isDisabled }) => {
    return {
      ...base,
      color: isDisabled ? '#4f4f4f' : isFocused ? '#fff' : '#4f4f4f',
      fontSize: '14px',
      lineHeight: '17px',
      fontFamily: 'Lato, sans-serif',
      backgroundColor: isDisabled ? '#bfbfbf' : isFocused ? '#69D5B1' : '#fff',
    };
  },
  multiValueLabel: base => ({
    ...base,
    backgroundColor: '#f2f2f2',
    color: '#4f4f4f',
    fontSize: '14px',
    lineHeight: '17px',
  }),
  multiValueRemove: base => ({
    ...base,
    backgroundColor: '#f2f2f2',
    color: '#4f4f4f',
    fontSize: '14px',
    lineHeight: '17px',
  }),
  indicatorSeparator: () => ({
    display: 'none',
  }),
};
