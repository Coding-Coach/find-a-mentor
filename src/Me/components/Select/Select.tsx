import { useContext, useLayoutEffect, useRef } from 'react';
import ReactSelect from 'react-select';
import type {Props} from 'react-select';
import { formFieldContext } from '../FormField/formContext';
import StateManager from 'react-select';

type SelectValue = {
  value: any;
  label: string;
}

type Option = SelectValue & {
  isDisabled?: boolean;
}

type OptionData = {name: string};

export type SelectProps = {
  name?: string;
  maxSelections?: number;
  options: Option[];
  isMulti?: boolean;
  value: SelectValue | SelectValue[];
  onChange: (value: SelectProps["value"], data: OptionData) => void;
}

const REACHED_MAX_OPTION: Option[] = [
  {
    label: 'Reached max items',
    value: undefined,
    isDisabled: true,
  },
];

export const Select = ({
  isMulti = false,
  value,
  onChange,
  maxSelections = Infinity,
  options,
  ...reactSelectOptions
}: SelectProps) => {
  const id = useContext(formFieldContext);
  const inputElRef = useRef<StateManager>(null);

  useLayoutEffect(() => {
    if (inputElRef.current !== null) {
      (inputElRef.current.select.inputRef as unknown as HTMLInputElement).autocomplete = 'nope';
    }
  }, []);

  return (
    <ReactSelect
      ref={inputElRef}
      isMulti={isMulti}
      menuPlacement="auto"
      hideSelectedOptions
      inputId={id}
      isSearchable
      value={value}
      onChange={onChange as any}
      options={
        isMulti && Array.isArray(value) && value.length >= maxSelections
          ? REACHED_MAX_OPTION
          : options
      }
      menuPortalTarget={document.body}
      styles={reactSelectStyles}
      {...reactSelectOptions}
    />
  );
};

const reactSelectStyles: Props["styles"] = {
  container: base => ({ ...base, width: '100%' }),
  control: base => ({
    ...base,
    fontSize: '14px',
    lineHeight: '17px',
    borderRadius: '3px',
    backgroundColor: '#fff',
    border: '1px solid #bfbfbf',
  }),
  placeholder: base => ({ ...base, color: '#898889' }),
  menuPortal: base => ({ ...base, zIndex: 1000 }),
  option: (base, { isFocused, isDisabled }) => {
    return {
      ...base,
      color: isDisabled ? '#4f4f4f' : isFocused ? '#fff' : '#4f4f4f',
      fontSize: '14px',
      lineHeight: '17px',
      fontFamily: 'Lato, sans-serif',
      backgroundColor: isDisabled ? '#bfbfbf' : isFocused ? '#69D5B1' : '#fff',
    };
  },
  multiValueLabel: base => ({
    ...base,
    backgroundColor: '#f2f2f2',
    color: '#4f4f4f',
    fontSize: '14px',
    lineHeight: '17px',
  }),
  multiValueRemove: base => ({
    ...base,
    backgroundColor: '#f2f2f2',
    color: '#4f4f4f',
    fontSize: '14px',
    lineHeight: '17px',
  }),
  indicatorSeparator: () => ({
    display: 'none',
  }),
};
