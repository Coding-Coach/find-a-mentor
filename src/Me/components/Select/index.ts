export { Select as default } from './Select';
