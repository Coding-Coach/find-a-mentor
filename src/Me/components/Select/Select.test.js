import React from 'react';
import Select from './';
import { render } from '@testing-library/react';

const options = [
  { value: 1, label: 'one' },
  { value: 2, label: 'two' },
  { value: 3, label: 'three' },
];
const noop = () => {};

describe('Select component', () => {
  it('renders properly', () => {
    const options = [
      { value: 1, label: 'one' },
      { value: 2, label: 'two' },
      { value: 3, label: 'three' },
    ];
    const { container } = render(
      <Select
        value={[]}
        onChange={noop}
        options={options}
        placeholder="Select a number"
        menuIsOpen
        menuPortalTarget={undefined}
      />
    );

    expect(container.firstChild).toMatchSnapshot();
  });
  it('does not allow to select more items than specified by maxSelections', () => {
    const maxItems = 2;

    const { getByText } = render(
      <Select
        isMulti={true}
        value={options.slice(0, maxItems)}
        onChange={noop}
        options={options}
        maxSelections={maxItems}
        placeholder="Select a number"
        menuIsOpen
        menuPortalTarget={undefined}
      />
    );

    expect(getByText('one')).toBeInTheDocument();
    expect(getByText('two')).toBeInTheDocument();
    expect(getByText('Reached max items')).toBeInTheDocument();
  });
});
