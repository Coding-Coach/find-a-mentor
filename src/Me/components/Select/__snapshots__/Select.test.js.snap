// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select component renders properly 1`] = `
<div
  class=" css-dyr6gj-container"
>
  <div
    class=" css-s29e3p-control"
  >
    <div
      class=" css-1hwfws3"
    >
      <div
        class=" css-gzyhiy-placeholder"
      >
        Select a number
      </div>
      <div
        class="css-b8ldur-Input"
      >
        <div
          class=""
          style="display: inline-block;"
        >
          <input
            aria-autocomplete="list"
            autocapitalize="none"
            autocomplete="nope"
            autocorrect="off"
            id="react-select-2-input"
            spellcheck="false"
            style="box-sizing: content-box; width: 2px; border: 0px; font-size: inherit; opacity: 1; outline: 0; padding: 0px;"
            tabindex="0"
            type="text"
            value=""
          />
          <div
            style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-size: inherit; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
          />
        </div>
      </div>
    </div>
    <div
      class=" css-1hb7zxy-IndicatorsContainer"
    >
      <span
        class=" css-1xjgjl1-IndicatorSeparator"
      />
      <div
        aria-hidden="true"
        class=" css-tlfecz-indicatorContainer"
      >
        <svg
          aria-hidden="true"
          class="css-6q0nyr-Svg"
          focusable="false"
          height="20"
          viewBox="0 0 20 20"
          width="20"
        >
          <path
            d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
          />
        </svg>
      </div>
    </div>
  </div>
  <div
    class=" css-26l3qy-menu"
  >
    <div
      class=" css-4ljt47-MenuList"
    >
      <div
        class=" css-zuk4g2-option"
        id="react-select-2-option-0"
        tabindex="-1"
      >
        one
      </div>
      <div
        class=" css-zuk4g2-option"
        id="react-select-2-option-1"
        tabindex="-1"
      >
        two
      </div>
      <div
        class=" css-zuk4g2-option"
        id="react-select-2-option-2"
        tabindex="-1"
      >
        three
      </div>
    </div>
  </div>
</div>
`;
